File: src/conversation/ConversationHandler.ts
// src/conversation/ConversationHandler.ts
import { LocalLlamaEngine } from '../engines/local/LocalLlamaEngine.js';
import { logger } from '../utils/logging.js';
import { ToolExecutor, ToolCall } from '../tools/ToolExecutor.js';
import { exec } from 'child_process';
import { promisify } from 'util';
import fs from 'fs';
import path from 'path';

const execAsync = promisify(exec);

export interface ConversationConfig {
  elevenLabsApiKey?: string;
  elevenLabsVoiceId?: string;
  enableTTS?: boolean;
  enableTools?: boolean;
}

export class ConversationHandler {
  private config: ConversationConfig;
  private isSpeaking = false;
  private toolExecutor: ToolExecutor;

  constructor(config?: ConversationConfig) {
    this.config = {
      elevenLabsApiKey: config?.elevenLabsApiKey || process.env.ELEVEN_API_KEY,
      elevenLabsVoiceId: config?.elevenLabsVoiceId || process.env.JARVIS_VOICE_ID || 'LE42bqYwZicKpZRastCO',
      enableTTS: config?.enableTTS ?? true,
      enableTools: config?.enableTools ?? true
    };

    this.toolExecutor = new ToolExecutor();

    if (this.config.enableTTS && !this.config.elevenLabsApiKey) {
      logger.warn('TTS enabled but ELEVEN_API_KEY not set - will use macOS say command as fallback');
    }
  }

  /**
   * Handle conversation from text input with reliable tool detection
   */
  async handleTextInput(
    text: string, 
    llmEngine: LocalLlamaEngine,
    sessionId: string = 'default'
  ): Promise<string> {
    try {
      // First, try direct tool detection (more reliable than LLM for 3B model)
      const directTool = this.detectDirectToolCall(text);
    
    if (directTool) {
      logger.info(`Direct tool detected: ${directTool.action}`);
      
      // Give immediate feedback
      if (this.config.enableTTS) {
        const feedbackMessage = this.getToolFeedbackMessage(directTool);
        if (feedbackMessage) {
          this.processAndSpeak(feedbackMessage, 'system').catch(() => {});
        }
      }
      
      const toolResult = await this.toolExecutor.execute(directTool);
      
      // Generate a natural response based on the tool result
      const finalResponse = await this.generateToolResultResponse(
        text, 
        toolResult, 
        llmEngine, 
        sessionId
      );
      
      await this.processAndSpeak(finalResponse, sessionId);
      return finalResponse;
    }

    // If no direct tool needed, use regular conversation
    const prompt = `You are JARVIS, a helpful AI assistant. You are concise and professional.
Never reference Tony Stark, Iron Man, or Marvel. Keep responses brief and natural.

User: ${text}
JARVIS:`;
    
    let response = '';
    
    for await (const event of llmEngine.generateStream(sessionId, prompt, {
      max_tokens: 150,
      temperature: 0.7
    })) {
      if (event.type === 'token' && event.text) {
        response += event.text;
      }
    }
    
    response = response.trim();
    await this.processAndSpeak(response, sessionId);
    return response;
    
  } catch (error) {
    logger.error('Conversation error:', error);
    const errorMessage = "I apologize, I encountered an error.";
    await this.processAndSpeak(errorMessage, sessionId);
    return errorMessage;
  }
}

/**
 * Detect direct tool calls using reliable patterns (3B model scope + basic interactions)
 */
private detectDirectToolCall(text: string): ToolCall | null {
  const lower = text.toLowerCase();
  
  // Time detection - 3B can handle this instantly
  if ((lower.includes('time') || lower.includes('clock')) && 
      (lower.includes('what') || lower.includes('current'))) {
    return { action: 'get_time', params: {} };
  }
  
  // Screen reading - 3B routes to vision API
  if (lower.includes('screen') && 
      (lower.includes('what') || lower.includes('on') || lower.includes('see'))) {
    return { action: 'read_screen', params: {} };
  }
  
  // Web search - 3B can handle this
  if (lower.includes('search') || lower.includes('google') || lower.includes('look up')) {
    const searchMatch = text.match(/(?:search|google|look up)\s+(?:for\s+)?(.+)/i);
    if (searchMatch) {
      return { action: 'web_search_and_open', params: { query: searchMatch[1] } };
    }
  }
  
  // App opening - 3B can handle this
  if (lower.includes('open') && !lower.includes('search')) {
    const appMatch = text.match(/open\s+(\w+)/i);
    if (appMatch) {
      return { action: 'open_app', params: { name: appMatch[1] } };
    }
  }
  
  // Basic interactions - 3B can route these
  
  // Clicking
  if (lower.includes('click')) {
    const clickMatch = text.match(/click\s+(?:on\s+)?(.+)/i);
    if (clickMatch) {
      return { action: 'click_element', params: { target: clickMatch[1] } };
    }
  }
  
  // Typing in specific places
  if (lower.includes('type') && (lower.includes('in') || lower.includes('into'))) {
    const typeInMatch = text.match(/type\s+(.+?)\s+(?:in|into)\s+(.+)/i);
    if (typeInMatch) {
      return { action: 'type_in_field', params: { text: typeInMatch[1], field: typeInMatch[2] } };
    }
  }
  
  // Simple typing
  if (lower.includes('type')) {
    const typeMatch = text.match(/type\s+(.+)/i);
    if (typeMatch) {
      return { action: 'type_text', params: { text: typeMatch[1] } };
    }
  }
  
  // Scrolling
  if (lower.includes('scroll')) {
    if (lower.includes('down')) {
      return { action: 'scroll', params: { direction: 'down' } };
    } else if (lower.includes('up')) {
      return { action: 'scroll', params: { direction: 'up' } };
    }
    return { action: 'scroll', params: { direction: 'down' } }; // default
  }
  
  // Press key combinations
  if (lower.includes('press')) {
    const keyMatch = text.match(/press\s+(.+)/i);
    if (keyMatch) {
      const keys = keyMatch[1].split(/\s+(?:and|plus|\+)\s+/);
      return { action: 'press_keys', params: { keys } };
    }
  }
  
  // Complex compound actions
  if (lower.includes('open') && lower.includes('and') && lower.includes('type')) {
    const openTypeMatch = text.match(/open\s+(\w+)\s+and\s+type\s+(.+)/i);
    if (openTypeMatch) {
      return { action: 'open_and_type', params: { app: openTypeMatch[1], text: openTypeMatch[2] } };
    }
  }
  
  // Math should route to 8B, so return null for that
  // Complex reasoning should route to 8B, so return null for that
  
  return null;
}

/**
 * Get contextual feedback message for tool execution (3B scope + interactions)
 */
private getToolFeedbackMessage(toolCall: ToolCall): string | null {
  switch (toolCall.action) {
    case 'get_time':
      return null; // Time is instant
    case 'open_app':
      return `One moment sir, opening ${toolCall.params.name}`;
    case 'read_screen':
      return "One moment sir, analyzing your screen";
    case 'web_search_and_open':
      return "One moment sir, opening browser and searching";
    case 'click_element':
      return `One moment sir, clicking on ${toolCall.params.target}`;
    case 'type_in_field':
      return `One moment sir, typing in ${toolCall.params.field}`;
    case 'type_text':
      return "One moment sir, typing text";
    case 'open_and_type':
      return `One moment sir, opening ${toolCall.params.app} and typing text`;
    case 'scroll':
      return null; // Scrolling is instant
    case 'press_keys':
      return null; // Key presses are instant
    default:
      return null;
  }
}

/**
 * Generate natural response based on tool result
 */
private async generateToolResultResponse(
  originalQuery: string,
  toolResult: any,
  engine: LocalLlamaEngine,
  sessionId: string
): Promise<string> {
  if (!toolResult.success) {
    return `I apologize, but I wasn't able to ${toolResult.action.replace('_', ' ')}. ${toolResult.error || 'Please try again.'}`;
  }

  // For some tools, just return the result directly
  if (toolResult.action === 'get_time' || toolResult.action === 'calculate') {
    return toolResult.output;
  }

  // For other tools, generate a natural response
  const prompt = `The user asked: "${originalQuery}"
I executed the action and got this result: ${toolResult.output}

Respond naturally as JARVIS, confirming what was done. Keep it brief:`;

  try {
    let response = '';
    for await (const event of engine.generateStream(sessionId + '_tool', prompt, {
      max_tokens: 50,
      temperature: 0.7
    })) {
      if (event.type === 'token' && event.text) {
        response += event.text;
      }
    }
    return response.trim();
  } catch (error) {
    // Fallback to simple confirmation
    return `Done. ${toolResult.output}`;
  }
}

/**
 * Process LLM response and convert to speech
 */
async processAndSpeak(response: string, sessionId: string): Promise<void> {
  if (!this.config.enableTTS) {
    logger.info(`JARVIS says: "${response}"`);
    return;
  }

  // Clean up response for speech
  const spokenText = this.cleanForSpeech(response);
  
  if (this.config.elevenLabsApiKey) {
    await this.speakWithElevenLabs(spokenText);
  } else {
    // Fallback to macOS say command
    await this.speakWithMacOS(spokenText);
  }
}

/**
 * Clean text for speech output
 */
private cleanForSpeech(text: string): string {
  // Remove special tokens
  text = text.replace(/<\|.*?\|>/g, '');
  text = text.replace(/\|im_end\|/g, '');
  text = text.replace(/\|im_start\|/g, '');
  text = text.replace(/assistant/g, '');
  
  // Remove markdown
  text = text.replace(/[*_~`]/g, '');
  
  // Remove code blocks
  text = text.replace(/```[\s\S]*?```/g, 'code block');
  
  // Remove URLs
  text = text.replace(/https?:\/\/[^\s]+/g, 'link');
  
  // Clean up multiple spaces and newlines
  text = text.replace(/\s+/g, ' ').trim();
  
  return text;
}

/**
 * Speak using ElevenLabs API
 */
private async speakWithElevenLabs(text: string): Promise<void> {
  if (this.isSpeaking) {
    logger.warn('Already speaking, skipping...');
    return;
  }

  this.isSpeaking = true;
  
  try {
    const url = `https://api.elevenlabs.io/v1/text-to-speech/${this.config.elevenLabsVoiceId}`;
    
    // Using fetch instead of axios since it's built into Node 18+
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Accept': 'audio/mpeg',
        'Content-Type': 'application/json',
        'xi-api-key': this.config.elevenLabsApiKey!
      },
      body: JSON.stringify({
        text,
        model_id: 'eleven_monolingual_v1',
        voice_settings: {
          stability: 0.6,
          similarity_boost: 1.0,
          style: 0.0,
          use_speaker_boost: true
        }
      })
    });

    if (!response.ok) {
      throw new Error(`ElevenLabs API error: ${response.status}`);
    }

    // Save audio to temp file
    const audioBuffer = await response.arrayBuffer();
    const tempFile = path.join('/tmp', `jarvis_${Date.now()}.mp3`);
    fs.writeFileSync(tempFile, Buffer.from(audioBuffer));
    
    // Play audio with increased volume
    await execAsync(`afplay -v 2 "${tempFile}"`);
    
    // Clean up
    fs.unlinkSync(tempFile);
    
    logger.info('TTS playback complete');
  } catch (error) {
    logger.error('ElevenLabs TTS error:', error);
    // Don't fallback - just log the error so we know what's wrong
    logger.error('TTS failed - check your ELEVEN_API_KEY');
  } finally {
    this.isSpeaking = false;
  }
}

/**
 * Fallback TTS using macOS say command
 */
private async speakWithMacOS(text: string): Promise<void> {
  if (this.isSpeaking) return;
  
  this.isSpeaking = true;
  
  try {
    // Use Daniel voice (British accent, closer to JARVIS)
    await execAsync(`say -v Daniel "${text.replace(/"/g, '\\"')}"`);
    logger.info('macOS TTS playback complete');
  } catch (error) {
    logger.error('macOS say command failed:', error);
  } finally {
    this.isSpeaking = false;
  }
}

/**
 * Stop any ongoing speech
 */
async stopSpeaking(): Promise<void> {
  if (this.isSpeaking) {
    try {
      // Kill any afplay or say processes
      await execAsync('pkill -f afplay || true');
      await execAsync('pkill -f say || true');
      this.isSpeaking = false;
    } catch (error) {
      logger.error('Error stopping speech:', error);
    }
  }
}
}

File: jarvis/conversation/conversation_handler.py
"""Ported from src/conversation/ConversationHandler.ts."""
from __future__ import annotations

import asyncio
import os
import re
import time
from dataclasses import dataclass
from typing import Optional

import httpx

from ..engines.local.local_llama_engine import GenerationParams, LocalLlamaEngine
from ..tools.tool_executor import ToolCall, ToolExecutor, ToolResult
from ..utils.logging import logger


async def _run(cmd: str) -> None:
    proc = await asyncio.create_subprocess_shell(cmd)
    await proc.wait()


@dataclass
class ConversationConfig:
    eleven_labs_api_key: Optional[str] = None
    eleven_labs_voice_id: Optional[str] = None
    enable_tts: bool = True
    enable_tools: bool = True


class ConversationHandler:
    def __init__(self, config: Optional[ConversationConfig] = None) -> None:
        self.config = config or ConversationConfig()
        self.config.eleven_labs_api_key = self.config.eleven_labs_api_key or os.environ.get("ELEVEN_API_KEY")
        self.config.eleven_labs_voice_id = self.config.eleven_labs_voice_id or os.environ.get("JARVIS_VOICE_ID", "LE42bqYwZicKpZRastCO")
        self.tool_executor = ToolExecutor()
        self._is_speaking = False

        if self.config.enable_tts and not self.config.eleven_labs_api_key:
            logger.warn("TTS enabled but ELEVEN_API_KEY not set - will use macOS say command as fallback")

    async def handle_text_input(self, text: str, llm_engine: LocalLlamaEngine, session_id: str = "default") -> str:
        try:
            direct_tool = self._detect_direct_tool_call(text)

            if direct_tool:
                logger.info(f"Direct tool detected: {direct_tool.action}")

                if self.config.enable_tts:
                    feedback = self._get_tool_feedback_message(direct_tool)
                    if feedback:
                        asyncio.create_task(self.process_and_speak(feedback, "system"))

                tool_result = await self.tool_executor.execute(direct_tool)
                final_response = await self._generate_tool_result_response(text, tool_result, llm_engine, session_id)
                await self.process_and_speak(final_response, session_id)
                return final_response

            prompt = (
                "You are JARVIS, a helpful AI assistant. You are concise and professional.\n"
                "Never reference Tony Stark, Iron Man, or Marvel. Keep responses brief and natural.\n\n"
                f"User: {text}\nJARVIS:"
            )

            response = ""
            async for event in llm_engine.generate_stream(session_id, prompt, GenerationParams(max_tokens=150, temperature=0.7)):
                if event.type == "token" and event.text:
                    response += event.text

            response = response.strip()
            await self.process_and_speak(response, session_id)
            return response
        except Exception as error:  # noqa: BLE001
            logger.error("Conversation error:", error)
            error_message = "I apologize, I encountered an error."
            await self.process_and_speak(error_message, session_id)
            return error_message

    def _detect_direct_tool_call(self, text: str) -> Optional[ToolCall]:
        lower = text.lower()

        if ("time" in lower or "clock" in lower) and ("what" in lower or "current" in lower):
            return ToolCall("get_time", {})

        if "screen" in lower and any(w in lower for w in ("what", "on", "see")):
            return ToolCall("read_screen", {})

        if any(w in lower for w in ("search", "google", "look up")):
            match = re.search(r"(?:search|google|look up)\s+(?:for\s+)?(.+)", text, re.IGNORECASE)
            if match:
                return ToolCall("web_search_and_open", {"query": match.group(1)})

        if "open" in lower and "search" not in lower:
            match = re.search(r"open\s+(\w+)", text, re.IGNORECASE)
            if match:
                return ToolCall("open_app", {"name": match.group(1)})

        if "click" in lower:
            match = re.search(r"click\s+(?:on\s+)?(.+)", text, re.IGNORECASE)
            if match:
                return ToolCall("click_element", {"target": match.group(1)})

        if "type" in lower and ("in" in lower or "into" in lower):
            match = re.search(r"type\s+(.+?)\s+(?:in|into)\s+(.+)", text, re.IGNORECASE)
            if match:
                return ToolCall("type_in_field", {"text": match.group(1), "field": match.group(2)})

        if "type" in lower:
            match = re.search(r"type\s+(.+)", text, re.IGNORECASE)
            if match:
                return ToolCall("type_text", {"text": match.group(1)})

        if "scroll" in lower:
            direction = "down" if "down" in lower else "up" if "up" in lower else "down"
            return ToolCall("scroll", {"direction": direction})

        if "press" in lower:
            match = re.search(r"press\s+(.+)", text, re.IGNORECASE)
            if match:
                keys = re.split(r"\s+(?:and|plus|\+)\s+", match.group(1))
                return ToolCall("press_keys", {"keys": keys})

        if "open" in lower and "and" in lower and "type" in lower:
            match = re.search(r"open\s+(\w+)\s+and\s+type\s+(.+)", text, re.IGNORECASE)
            if match:
                return ToolCall("open_and_type", {"app": match.group(1), "text": match.group(2)})

        return None

    def _get_tool_feedback_message(self, call: ToolCall) -> Optional[str]:
        messages = {
            "open_app": f"One moment sir, opening {call.params.get('name')}",
            "read_screen": "One moment sir, analyzing your screen",
            "web_search_and_open": "One moment sir, opening browser and searching",
            "click_element": f"One moment sir, clicking on {call.params.get('target')}",
            "type_in_field": f"One moment sir, typing in {call.params.get('field')}",
            "type_text": "One moment sir, typing text",
            "open_and_type": f"One moment sir, opening {call.params.get('app')} and typing text",
        }
        return messages.get(call.action)

    async def _generate_tool_result_response(
        self, original_query: str, tool_result: ToolResult, engine: LocalLlamaEngine, session_id: str
    ) -> str:
        if not tool_result.success:
            return f"I apologize, but I wasn't able to {tool_result.action.replace('_', ' ')}. {tool_result.error or 'Please try again.'}"

        if tool_result.action in ("get_time", "calculate"):
            return tool_result.output

        prompt = (
            f'The user asked: "{original_query}"\n'
            f"I executed the action and got this result: {tool_result.output}\n\n"
            "Respond naturally as JARVIS, confirming what was done. Keep it brief:"
        )

        try:
            response = ""
            async for event in engine.generate_stream(session_id + "_tool", prompt, GenerationParams(max_tokens=50, temperature=0.7)):
                if event.type == "token" and event.text:
                    response += event.text
            return response.strip()
        except Exception:  # noqa: BLE001
            return f"Done. {tool_result.output}"

    async def process_and_speak(self, response: str, session_id: str) -> None:
        if not self.config.enable_tts:
            logger.info(f'JARVIS says: "{response}"')
            return

        spoken_text = self._clean_for_speech(response)

        if self.config.eleven_labs_api_key:
            await self._speak_with_eleven_labs(spoken_text)
        else:
            await self._speak_with_macos(spoken_text)

    def _clean_for_speech(self, text: str) -> str:
        text = re.sub(r"<\|.*?\|>", "", text)
        text = text.replace("|im_end|", "").replace("|im_start|", "").replace("assistant", "")
        text = re.sub(r"[*_~`]", "", text)
        text = re.sub(r"```[\s\S]*?```", "code block", text)
        text = re.sub(r"https?://\S+", "link", text)
        return re.sub(r"\s+", " ", text).strip()

    async def _speak_with_eleven_labs(self, text: str) -> None:
        if self._is_speaking:
            logger.warn("Already speaking, skipping...")
            return

        self._is_speaking = True
        try:
            url = f"https://api.elevenlabs.io/v1/text-to-speech/{self.config.eleven_labs_voice_id}"
            async with httpx.AsyncClient(timeout=30) as client:
                response = await client.post(
                    url,
                    headers={"Accept": "audio/mpeg", "Content-Type": "application/json", "xi-api-key": self.config.eleven_labs_api_key},
                    json={
                        "text": text, "model_id": "eleven_monolingual_v1",
                        "voice_settings": {"stability": 0.6, "similarity_boost": 1.0, "style": 0.0, "use_speaker_boost": True},
                    },
                )
            if response.status_code != 200:
                raise RuntimeError(f"ElevenLabs API error: {response.status_code}")

            temp_file = f"/tmp/jarvis_{int(time.time() * 1000)}.mp3"
            with open(temp_file, "wb") as f:
                f.write(response.content)

            await _run(f'afplay -v 2 "{temp_file}"')
            os.remove(temp_file)
            logger.info("TTS playback complete")
        except Exception as error:  # noqa: BLE001
            logger.error("ElevenLabs TTS error:", error)
            logger.error("TTS failed - check your ELEVEN_API_KEY")
        finally:
            self._is_speaking = False

    async def _speak_with_macos(self, text: str) -> None:
        if self._is_speaking:
            return
        self._is_speaking = True
        try:
            escaped = text.replace('"', '\\"')
            await _run(f'say -v Daniel "{escaped}"')
            logger.info("macOS TTS playback complete")
        except Exception as error:  # noqa: BLE001
            logger.error("macOS say command failed:", error)
        finally:
            self._is_speaking = False

    async def stop_speaking(self) -> None:
        if self._is_speaking:
            try:
                await _run("pkill -f afplay || true")
                await _run("pkill -f say || true")
                self._is_speaking = False
            except Exception as error:  # noqa: BLE001
                logger.error("Error stopping speech:", error)
